src/main/java/com/tabnine/statusBar/StatusBarUpdater.kt

```diff
@@ -0,0 +1,94 @@
+package com.tabnine.statusBar
+
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.project.ProjectManager
+import com.intellij.openapi.wm.WindowManager
+import com.tabnine.binary.BinaryRequestFacade
+import com.tabnine.binary.requests.statusBar.StatusBarPromotionBinaryRequest
+import com.tabnine.binary.requests.statusBar.StatusBarPromotionBinaryResponse
+import com.tabnine.binary.requests.statusBar.StatusBarPromotionShownRequest
+import java.util.concurrent.atomic.AtomicBoolean
+
+class StatusBarUpdater(private val binaryRequestFacade: BinaryRequestFacade) {
+
+    private companion object {
+        val synchronizer = AtomicBoolean()
+    }
+
+    fun updateStatusBar() {
+        ApplicationManager.getApplication().executeOnPooledThread { requestStatusBarMessage() }
+    }
+
+    fun requestStatusBarMessage() {
+        try {
+            if (!synchronizer.compareAndSet(false, true)) {
+                return // abort if there's already a request in progress
+            }
+            val statusBarPromotionWidgets = getStatusBarsWidgets()
+            if (statusBarPromotionWidgets.isEmpty()) {
+                return
+            }
+
+            val promotion = binaryRequestFacade.executeRequest(StatusBarPromotionBinaryRequest())
+
+            if (promotion != null) {
+                updateStatusBars(statusBarPromotionWidgets, promotion)
+                binaryRequestFacade.executeRequest(
+                    StatusBarPromotionShownRequest(
+                        promotion.id,
+                        promotion.message ?: "undefined", promotion.notificationType,
```

Comment:
extract `"undefined"` to a const with a meaningful name, it looks like a bug here (I know it is not)

Reply:
Why do we even pass "undefined"? I might set it to 'undefined' here while another plugin will set it to 'empty' etc'.
We should return null and let the binary replace this with whatever 'empty' value it wants.

Reply:
🤷